Post introducing the SSG Test Suite throught the approachable path (#19)
ComplianceAsCode/ComplianceAsCode.github.io: _posts/2021-03-15-tests_howto.md

> Testing of rules and remediations

Reviewer: ```suggestion
Testing rules and remediations
```

=================================

How do we make sure that rules in the ComplianceAsCode repository are correct?
Before we answer that, let's take a look at the big picture.
The part of a rule that is a code of some kind can include OVAL definition, a Bash snippet, and a set of Ansible tasks.

OVAL is not a very expressive language, and it is not very readable either.
Bash is a shell in which you can write simple programs.
Robustness of those programs and readability of the code are not a strong aspects either.
Finally, Ansible is primarily a configuration management software, and hardening is a more delicate use case, so the readability of a rule's Ansible content can suffer from it.
So being sure that a given rule is written correctly is not trivial.

However, we are not finished here.
The ComplianceAsCode repository hosts more than one thousand rules.
Many of them are templated, i.e. they are slight variations of one principle, but there are still hundreds that are crafted by hand.
Therefore, making sure that even rules that are used by a given product are correct is tough to scale - that means verifying tens to hundreds of rules.

So how do we do it in the project?

> As you can learn very quickly by visiting the [tests](https://github.com/ComplianceAsCode/content/tree/master/tests#readme) folder, the project features a test suite.

Reviewer: I got the idea and I appreciate your creativity but I somehow don't like it - the text talks about a folder, but after clicking on a link you are shown the readme, iow after clicking the reader gets a different thing than he expects. I suggest splitting this into 2 links.

```suggestion
As you can learn very quickly by visiting the [tests](https://github.com/ComplianceAsCode/content/tree/master/tests) folder, the project features a test suite.
The test suite is described in [README](https://github.com/ComplianceAsCode/content/blob/master/tests/README.md) located in this directory.
```

Reviewer 2: incorporated

The Test Suite
--------------

What does the test suite do?
It is able to verify that a check passes or fails as expected.
Moreover, in case that if it fails, it tests that a remediation is able to fix the system, so the check on the fixed system passes.

It does so by using virtualization (containers or VMs) to set up the virtualized system.
You can see the `tests` directory in rule with scripts - each script represents one test scenario, which sets up the system to a certain state that should be picked up by the scanner.
A test scenario may be tied to a particular product or profile, but if the rule is about correct contents of a configuration file, such scenario can be applied to any product.

The test suite executes the scenario script, and then it performs the scan.
If the scanner reports a failed rule, and the scenario was supposed to fail, remediation is applied, and another scan is conducted.
Results are then reported to the command-line, and saved to the log file.

Therefore, the test suite performs these tasks:

1. Environment preparation.
2. Initial scan - this can be the last thing if the test scenario is supposed to pass.

> 3. Remediation - in case of `oscap` remediation, this step includes the final scan as well.

Reviewer 2: AFAIK, It should be the same for `ansible` and `bash`. In case of ansible you need ansible package installed.

4. Scan of remediated system.

Testing as an unprivileged user
-------------------------------

Since `podman` became a standard part of Linux distributions, one can build and run containers locally without additional privileges, so let's see how such test can look like.

> First of all, we will need a test environment.

Reviewer: Do I have to have some packages installed on my machine before I can run this?

Reviewer 2: I think for container backend you only need podman to be installed.

Reviewer 2: only podman for container testing, it should be already installed (the post is assuming that I guess)

> The project already contains predefined recipes, so let's leverage them, and build the test container:

Reviewer: Recipes? I think you meant "tools to set up test environment".

Reviewer 2: accepted the suggestion with a slight change

```shell
$ tests/build_test_container.sh
```

That's it, the script builds a Fedora-based test container image called `ssg_test_suite` by default, which can get us quite far, as you will see.
The script offers options to customize its behavior - simply call `tests/build_test_container.sh -h` to see how.
If the selection of flavors caught your eye, those have to be in line with contents of the [Dockerfiles](https://github.com/ComplianceAsCode/content/tree/master/Dockerfiles) directory, specifically to `test_suite-...` recipes.
However, you can use a Fedora-based image to test many rules even if you are a Debian content developer.

Next, we build the content that we would like to test - so let's dare and test a rule in the RHEL7 context.
We do it to demonstrate the low sensitivity of the tested content to its environment - RHEL7 is surely not Fedora.
You can go ahead and use a different product further from Fedora, just make sure that its benchmark includes the `accounts_tmout` rule that we are going to use as an example.

```shell
$ ./build_product rhel7
```

> ... and we have the datastream!

Reviewer: ```suggestion
... and we have the data stream!
```

Reviewer 2: I don't think we need to change all the occurrences of `datastream`

Now let's test a rule that is text-based, cross-platform and easy to follow - the famous [accounts_tmout](https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/accounts/accounts-session/accounts_tmout/rule.yml) one:

```shell

> $ tests/test_rule_in_container.sh --dontclean accounts_tmout

Reviewer: `--dontclean` doesn't exist

...
INFO - Script line_not_there.fail.sh using profile (all) OK
INFO - Script multiline.fail.sh using profile (all) OK
INFO - Script wrong_value.fail.sh using profile (all) OK
INFO - Script correct_value.pass.sh using profile (all) OK
```

...

So what happened?
The test suite executed `oscap` in the Fedora container, but two things had to be adjusted to make it work:

> 1. The applicability of the datastream is extended to include the system used for the actual test - in this case Fedora.

Reviewer: Do you recommend testing data streams for other platforms on different platforms? Isn't testing on the same platform the best practice? I assume you will get worng results in case of rules that differ between platforms.

Can you add more text about this topic?

Reviewer 2: This was the only thing I have not touched. I don't think it should be a blocker for the blog post. I think it was clear enough that this is just to make starting of tests easier and the appropriate machine/container should be used to serious testing.

> The content still corresponds to the system that the datastream has been compiled for, in this case that would be the RHEL7 content, but it could be SLE or Ubuntu content as well.

Reviewer: ```suggestion
The content still corresponds to the system that the data stream has been compiled for, in this case that would be the RHEL7 content, but it could be SLE or Ubuntu content as well.
```

2. Our example rule is marked as machine-only, i.e. not applicable in container environment, which would the test of the rule in a container impossible.
The test suite removes this applicability restriction from the rule and its remediations.
Some rules don't make sense in a container in production, but technically, they can be formally tested and remediated the same way as they would be on a bare-metal system.

That sounds like the wrapper did a bit of a black magic, but you can always pass the `--dry-run` argument to reveal what it did:

```shell

> $ tests/test_rule_in_container.sh --dry-run --dontclean accounts_tmout

Reviewer: `--dry-run --dontclean` doesn't exist

Reviewer: Aha I can see there is a PR. https://github.com/ComplianceAsCode/content/pull/6675/files

Consider, for instance, [comment.fail.sh](https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/accounts/accounts-session/accounts_tmout/tests/comment.fail.sh)

There are following things that we can tell about that scenario:

- Its filename ends by `fail.sh` - that means that the scenario is supposed to make the rule scan to fail.
Therefore, this scenario tests the ability of the OVAL to detect that the system is broken, and the ability of the remediation to fix it.
Precisely, we test the ability of the remediation to satisfy the OVAL check - the test suite doesn't do functional tests, i.e. it doesn't check that the system behaves in a particular way - it just executes the scanner.
- Inside, we see that there is a line containing `# variables = var_accounts_tmout=600` - that indicates that the rule is parametrized, and we force the value of the parameter regardless of the rule's default, or of any possible profile setting.
[More metadata](https://github.com/ComplianceAsCode/content/tree/master/tests#scenarios-format) are supported.

> Let's look a completely different scenario [missing.fail.sh](https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/bootloader-grub2/non-uefi/grub2_password/tests/missing.fail.sh) - for the [grub2_password](https://github.com/ComplianceAsCode/content/tree/master/linux_os/guide/system/bootloader-grub2/non-uefi/grub2_password) rule - let's check its contents:

Reviewer: ```suggestion
Let's look at a completely different scenario [missing.fail.sh](https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/bootloader-grub2/non-uefi/grub2_password/tests/missing.fail.sh) - for the [grub2_password](https://github.com/ComplianceAsCode/content/tree/master/linux_os/guide/system/bootloader-grub2/non-uefi/grub2_password) rule - and check its contents:
```

Reviewer 2: done

- There is a `# remediation = none` metadata directive - that one is useful if the rule doesn't have a remediation.
In that case, the test suite is satisfied enough just after the expected failure occurs.
A remediation not being available is usually an exception, not an intention, and it can "disappear" as an unwanted side-effect of a metadata update, or because of a build system glitch.
Therefore, the test suite has the opt-out approach for remediations.
- Finally, there is the `. $SHARED/grub2.sh` shell sourcing.
The scenario setup code [can be shared](https://github.com/ComplianceAsCode/content/tree/master/tests#sharing-code-among-test-scenarios) to any scenario, so there is no need for code duplication.

I believe that now you know enough to start experimenting with test scenarios - keep in mind that the test suite can run them selectively by means of the `--scenarios` argument, which is handy when only one scenario causes issues..

Conclusion
----------

The blog post grew quite long, but we just scratched the surface - the test suite can do much more!

> It can test profiles, when it can determine how "green" can the datastream make typically a freshly installed system.

Reviewer: ```suggestion
It can test profiles, when it can determine how "green" can the data stream make typically a freshly installed system.
```

This type of test just runs the scan with remediations if needed (i.e. no scenarios), but it can reveal conflicts between rules that are selected by the profile.
The test suite can also operate in combined mode that combines the profile test with rule tests - in that case, every rule in a profile is tested thoroughly using its test scenarios.

Although the container backend is very convenient, it's not the best representation of a system.

> A virtual machine feels much closer to the bare-metal, and the test suite supports the `libvirt` backend, which leverages all those snapshots and VM control capabilities of `libvirt` to execute tests.

Reviewer: ```suggestion
A virtual machine feels much closer to the bare-metal, and the test suite supports the `libvirt` backend which leverages VM control capabilities of `libvirt` to execute tests.
```

Reviewer 2: incorporated
